compiler/ir/backend.js/src/org/jetbrains/kotlin/ir/backend/js/lower/EnumClassLowering.kt

```diff
@@ -516,26 +484,12 @@ class EnumSyntheticFunctionsLowering(val context: JsIrBackendContext): Declarati
         }
     }
 
-    private fun List<IrExpression>.toArrayLiteral(arrayType: IrType, elementType: IrType): IrExpression {
-        val irVararg = IrVarargImpl(UNDEFINED_OFFSET, UNDEFINED_OFFSET, arrayType, elementType, this)
-
-        return IrCallImpl(
-            UNDEFINED_OFFSET, UNDEFINED_OFFSET, arrayType,
-            context.intrinsics.arrayLiteral,
-            typeArgumentsCount = 0,
-            valueArgumentsCount = 1
-        ).apply {
-            putValueArgument(0, irVararg)
-        }
-    }
-
     private fun createEnumValuesBody(valuesFun: IrFunction, irClass: IrClass): IrBlockBody {
-        val backendContext = context
         return context.createIrBuilder(valuesFun.symbol).run {
             irBlockBody {
+                val instances = irClass.enumEntries.map { irCall(it.getInstanceFun!!) }
                 +irReturn(
-                    irClass.enumEntries.map { irCall(it.getInstanceFun!!) }
-                        .toJsArrayLiteral(backendContext, valuesFun.returnType, irClass.defaultType)
+                    IrVarargImpl(UNDEFINED_OFFSET, UNDEFINED_OFFSET, valuesFun.returnType, irClass.defaultType, instances)
```

Comment:
feels like an abuse

Reply:
We used to have the same IrVararg node in "arrayOf(..)" call anyway. Now we have it without useless `arrayOf`

Reply:
Anyway, I think it's invalid IR. Depending on context vararg could be translated differently.

Reply:
What kind of different meaning can IrVararg have?

If it is invalid IR, then we get invalid IR from `arrayOf` calls:

Given Kotlin code
```
    intArrayOf(1, 2, 3)
    arrayOf("A", 1, 1 + 2)
``` 


We have this IR after function inlining:

```
        TYPE_OP type=kotlin.Unit origin=IMPLICIT_COERCION_TO_UNIT typeOperand=kotlin.Unit
          BLOCK type=kotlin.IntArray origin=null
            VAR IR_TEMPORARY_VARIABLE name:tmp0_foo_0 type:kotlin.IntArray [val]
              BLOCK type=kotlin.IntArray origin=(InlinerExpressionLocationHint : FUN name:box visibility:public modality:FINAL <> () returnType:kotlin.String @null)
                VARARG type=kotlin.IntArray varargElementType=kotlin.Int
                  CONST Int type=kotlin.Int value=1
                  CONST Int type=kotlin.Int value=2
                  CONST Int type=kotlin.Int value=3
            RETURN type=kotlin.Nothing from='BLOCK type=kotlin.IntArray origin=null'
              GET_VAR 'val tmp0_foo_0: kotlin.IntArray [val] declared in foo.box' type=kotlin.IntArray origin=null
        TYPE_OP type=kotlin.Unit origin=IMPLICIT_COERCION_TO_UNIT typeOperand=kotlin.Unit
          BLOCK type=kotlin.Array<out kotlin.Any> origin=null
            VAR IR_TEMPORARY_VARIABLE name:tmp1_foo2_0 type:kotlin.Array<out kotlin.Any> [val]
              BLOCK type=kotlin.Array<out kotlin.Any> origin=(InlinerExpressionLocationHint : FUN name:box visibility:public modality:FINAL <> () returnType:kotlin.String @null)
                VARARG type=kotlin.Array<out kotlin.Any> varargElementType=kotlin.Any
                  CONST String type=kotlin.String value="A"
                  CONST Int type=kotlin.Int value=1
                  CALL 'public final fun plus (other: kotlin.Int): kotlin.Int [operator] declared in kotlin.Int' type=kotlin.Int origin=PLUS
                    $this: CONST Int type=kotlin.Int value=1
                    other: CONST Int type=kotlin.Int value=2
```

Reply:
good point, so looks like it should be called something like "IrArrayLiteral" :)